server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java

```diff
@@ -83,13 +82,26 @@ default Stream<GroupModel> searchForGroupByNameStream(RealmModel realm, String s
      * This is done until the group node does not have a parent (root group)
      *
      * @param realm Realm.
-     * @param search Case sensitive searched string.
-     * @param exact Boolean which defines wheather search param should be matched exactly.
+     * @param search searched string.
+     * @param exact Boolean which defines wheather search param should be matched exactly (including case).
      * @param firstResult First result to return. Ignored if negative or {@code null}.
      * @param maxResults Maximum number of results to return. Ignored if negative or {@code null}.
      * @return Stream of root groups that have the given string in their name themself or a group in their child-collection has.
      * The returned hierarchy contains siblings that do not necessarily have a matching name. Never returns {@code null}.
      */
     Stream<GroupModel> searchForGroupByNameStream(RealmModel realm, String search, Boolean exact, Integer firstResult, Integer maxResults);
 
+    /**
+     * Returns the subgroups of a given group id
+     * @param realm Realm
+     * @param id the id of parent group to be used to load subgroups with
+     * @param firstResult First result to return, Ignored if negative or {@code null}.
+     * @param maxResults Maximum number of results to return. Ignored if negative or {@code null}.
+     * @return A Stream of groups that all have the given parent group as their parent.
+     */
+    default Stream<GroupModel> searchForSubgroupsByParentIdStream(RealmModel realm, String id, Integer firstResult, Integer maxResults) {
```

Comment:
Why are these two methods added here? Isn't it enough to add these only to `GroupModel`?

Reply:
I was wondering, whether this can be added only to `GroupModel` not `GroupLookupProvider` nor `GroupProvider`?

Reply:
For me, it's really confusing to have some of the group lookup functionality on the `GroupProvider` and some on the `GroupModel`. 

I do understand the appeal of being able to use the `GroupModel` to directly get the related subgroups, rather than having to go back and call the session; the implementation of that behavior on the adapters rather than the provider implementations seems to go against the pattern of how database interactions are otherwise encapsulated though.

That's my thought process there but ultimately if it needs to be moved back I don't mind doing it!

Reply:
I understand your confusion. In general, `GroupProvider` is something that goes to the storage search for some groups based on some parameters and returns it to the caller. While `GroupModel` represents one group returned by the `GroupProvider`.

Now, there is a question about where to put this method `searchForSubgroupsByParentIdNameStream`. It can be added to both. Personally, I prefer to have it only in one of them as I don't like it when there are two ways of doing the same thing. The API is then really confusing. In Keycloak we have many methods that are doing nothing just calling another method. This is something we should get rid of and the best way to start is not to include such methods anymore.

My reasoning behind `GroupModel` is:
- We already have [`getSubGroupsStream`](https://github.com/keycloak/keycloak/blob/main/server-spi/src/main/java/org/keycloak/models/GroupModel.java#L111) so it is consistent with what we currently have
- Some storages may benefit from knowing more details about the parent group than ID. For some implementations, the current method may look like as follows 1. find group by Id 2. search child groups, which is not very effective as the parent group was most probably already loaded before we called this method.

If you decide you still want to include this in `GroupLookupProvider` it is ok but we should then make it consistent and remove the methods from `GroupModel`.